cpmpy/solvers/choco.py

```diff
@@ -0,0 +1,585 @@
+            are auxiliary variables created by transformations.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: self
+        """
+        # add new user vars to the set
+        get_variables(cpm_expr, collect=self.user_vars)
+
+        # transform and post the constraints
+        for con in self.transform(cpm_expr):
+            c = self._get_constraint(con)
+            if c is not None:
+                c.post()
+
+        return self
+
+    def _get_constraint(self, cpm_expr):
+        """
+        Get a solver's constraint by a supported CPMpy constraint
+
+        :param cpm_expr: CPMpy expression
+        :type cpm_expr: Expression
+
+        """
+        from pychoco.variables.intvar import IntVar
+
+        # Operators: base (bool), lhs=numexpr, lhs|rhs=boolexpr (reified ->)
+        if isinstance(cpm_expr, Operator):
+            # 'and'/n, 'or'/n, '->'/2
+            if cpm_expr.name == 'and':
+                return self.chc_model.and_(self.solver_vars(cpm_expr.args))
+            elif cpm_expr.name == 'or':
+                return self.chc_model.or_(self.solver_vars(cpm_expr.args))
+            elif cpm_expr.name == '->':
+                cond, subexpr = cpm_expr.args
+                assert isinstance(cond, _BoolVarImpl), f"Implication constraint {cpm_expr} must have BoolVar as lhs"
+                lhs = self.solver_var(cond) # should always be boolvar due to only_bv_reifies
+                # right hand side
+                if isinstance(subexpr, _BoolVarImpl):
+                    # bv -> bv
+                    # PyChoco does not have "implies" constraint
+                    return self.chc_model.or_([~lhs, self.solver_var(subexpr)])
+                else:
+                    # bv -> boolexpr
+                    # the `reify_rewrite()` transformation ensures that only reifiable rhs remain here
+                    if subexpr.name == 'not':
+                        bv = self._get_constraint(subexpr.args[0]).reify()
+                        return self.chc_model.or_([~lhs, ~bv])
+                    else:
+                        bv = self._get_constraint(cpm_expr.args[1]).reify()
+                        return self.chc_model.or_([~lhs, bv])
+            else:
+                raise NotImplementedError("Not a known supported Choco Operator '{}' {}".format(
+                    cpm_expr.name, cpm_expr))
+
+        # Comparisons: both numeric and boolean ones
+        # numexpr `comp` bvar|const
+        # Choco accepts only int32, not int64
+        elif isinstance(cpm_expr, Comparison):
+            for i in range(len(cpm_expr.args)):
+                if isinstance(cpm_expr.args[i], np.integer):
+                    cpm_expr.args[i] = int(cpm_expr.args[i])
+            lhs, rhs = cpm_expr.args
+
+            if lhs.is_bool() and rhs.is_bool(): #boolean equality -- Reification
```

Comment:
I noticed that canonical comparison can output things that would be normalised away in simplify_bool such as `bv > 0`, but since it is now again after flatten they can appear here.
So we need to either use is_boolexpr(lhs) and is_boolexpr(rhs) here (which will treat bv > 0 as a numexpr, but that seems to work). Or change canonical comparison to output bv in stead of bv > 0 (might be harder i guess)

Reply:
i.e _canonical_comparison(sum(bv) > 0)_ outputs` bv > 0 `

Reply:
Yes, but simplify bool does not simplify away sum(bv) >0, and then canonical comparison outputs bv > 0 which gives an error when using .is_bool.
I guess you could argue that's a bug in simplify_bool actually but leaving .is_bool would require you to be very sure to cover all cases

Reply:
I see